### PR TITLE
USSD API implementation inbound/subscriptions

### PR DESCRIPTION
**Pull request by cjpannila**

### CODE REVIEW

File: com.wso2telco.dep.hub.ussdapi/com.wso2telco.dep.hub.ussdapi.synapse/src/main/synapse-config/sequences/com.wso2telco.dep.hub.ussdapi.resource.mo.subscribe.notification.splitter.Sequence.xml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                <script language="js"><![CDATA[var payload = mc.getPayloadXML();
+                delete payload..*::operatorCode;
+                delete payload..*::shortCodes;
+                mc.setPayloadXML(payload);]]></script>
+                <!-- retrieve operator's endpoint -->
+                <sequence key="com.wso2telco.dep.common.operator.endpoint.retriever.Sequence"/>
+                <filter regex="true" source="get-property('default', 'ENDPOINT_NOT_PROVISIONED')">
+                    <then>
+                        <property name="messageStatus" scope="default" type="STRING" value="Not Provisioned"/>
+                        <!-- dummy endpoint call to step-out from iterate -->
```
Comment (Buddhima) [inline com.wso2telco.dep.hub.ussdapi/com.wso2telco.dep.hub.ussdapi.synapse/src/main/synapse-config/sequences/com.wso2telco.dep.hub.ussdapi.resource.mo.subscribe.notification.splitter.Sequence.xml]: better to indicate with a "warn" log

File: com.wso2telco.dep.hub.ussdapi/com.wso2telco.dep.hub.ussdapi.synapse/src/main/synapse-config/sequences/com.wso2telco.dep.hub.ussdapi.resource.mo.subscribe.notification.splitter.fault.Sequence.xml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<sequence name="com.wso2telco.dep.hub.ussdapi.resource.mo.subscribe.notification.splitter.fault.Sequence" trace="disable" xmlns="http://ws.apache.org/ns/synapse">
+    <filter regex="Not Provisioned" source="$ctx:messageStatus">
```
Comment (Buddhima) [inline com.wso2telco.dep.hub.ussdapi/com.wso2telco.dep.hub.ussdapi.synapse/src/main/synapse-config/sequences/com.wso2telco.dep.hub.ussdapi.resource.mo.subscribe.notification.splitter.fault.Sequence.xml:3]: add warn / error logs

File: com.wso2telco.dep.hub.ussdapi/com.wso2telco.dep.hub.ussdapi.synapse/src/main/synapse-config/sequences/com.wso2telco.dep.hub.ussdapi.resource.mo.subscribe.notification.gather.Sequence.xml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    </enrich>
+    <enrich>
+        <source clone="true" property="messageStatus" type="property"/>
+        <target xpath="//shortCodes/status/text()"/>
+    </enrich>
+    <property name="collection" scope="default">
+        <subscription xmlns=""/>
+    </property>
+    <aggregate id="ussd.subscribe.notifications.splitter">
+        <completeCondition>
```
Comment (Buddhima) [inline com.wso2telco.dep.hub.ussdapi/com.wso2telco.dep.hub.ussdapi.synapse/src/main/synapse-config/sequences/com.wso2telco.dep.hub.ussdapi.resource.mo.subscribe.notification.gather.Sequence.xml]: add a timeout for aggregation

File: com.wso2telco.dep.hub.ussdapi/com.wso2telco.dep.hub.ussdapi.synapse/src/main/synapse-config/templates/com.wso2telco.dep.hub.ussdapi.notification.url.generating.Template.xml [hunk 4]
```diff
@@ -0,0 +1,60 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<template name="com.wso2telco.dep.hub.ussdapi.notification.url.generating.Template" xmlns="http://ws.apache.org/ns/synapse">
+    <parameter name="apiContext"/>
```
Comment (Buddhima) [inline com.wso2telco.dep.hub.ussdapi/com.wso2telco.dep.hub.ussdapi.synapse/src/main/synapse-config/templates/com.wso2telco.dep.hub.ussdapi.notification.url.generating.Template.xml]: pls check if we can reuse common template here

**Comment by cjpannila** [comment]
Related pull request in mediation-dep-common: https://github.com/WSO2Telco/mediation-dep-common/pull/112

File: com.wso2telco.dep.hub.ussdapi/com.wso2telco.dep.hub.ussdapi.synapse/src/main/synapse-config/templates/com.wso2telco.dep.hub.ussdapi.notification.url.generating.Template.xml (hunk 4 above)
Comment (Buddhima) [reply com.wso2telco.dep.hub.ussdapi/com.wso2telco.dep.hub.ussdapi.synapse/src/main/synapse-config/templates/com.wso2telco.dep.hub.ussdapi.notification.url.generating.Template.xml]: now this file can be deleted

File: com.wso2telco.dep.hub.ussdapi/com.wso2telco.dep.hub.ussdapi.synapse/src/main/synapse-config/sequences/com.wso2telco.dep.hub.ussdapi.resource.mo.subscribe.notification.splitter.Sequence.xml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                <script language="js"><![CDATA[var payload = mc.getPayloadXML();
+                delete payload..*::operatorCode;
+                delete payload..*::shortCodes;
+                mc.setPayloadXML(payload);]]></script>
+                <!-- retrieve operator's endpoint -->
+                <sequence key="com.wso2telco.dep.common.operator.endpoint.retriever.Sequence"/>
+                <filter regex="true" source="get-property('default', 'ENDPOINT_NOT_PROVISIONED')">
+                    <then>
+                    	<log level="custom" category="ERROR">
+							<property name="ERROR" value="Operator endpoint not provisioned"/>
```
Comment (Buddhima) [inline com.wso2telco.dep.hub.ussdapi/com.wso2telco.dep.hub.ussdapi.synapse/src/main/synapse-config/sequences/com.wso2telco.dep.hub.ussdapi.resource.mo.subscribe.notification.splitter.Sequence.xml]: better to add operator name as well

File: com.wso2telco.dep.hub.ussdapi/com.wso2telco.dep.hub.ussdapi.synapse/src/main/synapse-config/sequences/com.wso2telco.dep.hub.ussdapi.resource.mo.subscribe.notification.splitter.fault.Sequence.xml (hunk 2 above)
Comment (Buddhima) [reply com.wso2telco.dep.hub.ussdapi/com.wso2telco.dep.hub.ussdapi.synapse/src/main/synapse-config/sequences/com.wso2telco.dep.hub.ussdapi.resource.mo.subscribe.notification.splitter.fault.Sequence.xml:3]: add operator name as well

**Pull request merged**